assets/js/googlesitekit/datastore/user/surveys.js

```diff
@@ -185,58 +157,66 @@ const baseActions = {
 		},
 		function* ( triggerID, options = {} ) {
 			const { ttl = 0 } = options;
-			const { select, dispatch, resolveSelect } =
-				yield commonActions.getRegistry();
-
-			// Wait for user authentication state to be available before selecting.
-			yield commonActions.await(
-				resolveSelect( CORE_USER ).getAuthentication()
-			);
-
-			if ( ! select( CORE_USER ).isAuthenticated() ) {
+			const { select, resolveSelect } = yield commonActions.getRegistry();
+			const {
+				isAuthenticated,
+				isSurveyTimedOut,
+				isSurveyTriggerLocked,
+				getSurveyTimeouts,
+			} = select( CORE_USER );
+
+			// The lock prevents multiple concurrent triggers for the same ID.
+			if ( isSurveyTriggerLocked( triggerID ) ) {
 				return {};
 			}
+			yield lockSurveyTrigger( triggerID );
 
-			// Await for surveys to be resolved before checking timeouts.
-			yield commonActions.await(
-				resolveSelect( CORE_USER ).getSurveyTimeouts()
-			);
+			try {
+				// Wait for user authentication state to be available before selecting.
+				yield commonActions.await(
+					resolveSelect( CORE_USER ).getAuthentication()
+				);
 
-			const isTimedOut =
-				select( CORE_USER ).isSurveyTimedOut( triggerID );
-			const isTimingOut = select( CORE_USER ).isTimingOutSurvey(
-				triggerID,
-				ttl
-			);
+				if ( ! isAuthenticated() ) {
+					return {};
+				}
+
+				yield commonActions.await(
+					resolveSelect( CORE_USER ).getSurveyTimeouts()
+				);
+
+				// Check if persistent timeout for this trigger is set
+				// or if an in-memory timeout is present (see below).
+				if ( isSurveyTimedOut( triggerID ) ) {
+					return { response: {}, error: false };
+				}
 
-			// Both isTimedOut and isTimingOut variables are already resolved since they depend on
-			// the getSurveyTimeouts selector which we've resolved just before getting these variables.
-			if ( ! isTimedOut && ! isTimingOut ) {
 				const { response, error } =
 					yield fetchTriggerSurveyStore.actions.fetchTriggerSurvey(
-						triggerID
+						triggerID,
+						ttl
 					);
 
 				if ( error ) {
 					return { response, error };
 				}
 
-				// If TTL isn't empty, then sleep for 30s and set survey timeout.
 				if ( ttl > 0 ) {
-					yield new Promise( ( resolve ) => {
-						setTimeout( resolve, 30000 );
-					} );
-
-					yield commonActions.await(
-						dispatch( CORE_USER ).setSurveyTimeout( triggerID, ttl )
+					// Survey timeouts should be available here; fallback to empty array in case of error.
+					const timeouts = getSurveyTimeouts() || [];
+					// Add the trigger to the list of timeouts directly (next time it will come from server side).
+					yield fetchGetSurveyTimeoutsStore.actions.receiveGetSurveyTimeouts(
+						[ ...timeouts, triggerID ]
 					);
 				}
-			}
 
-			return {
-				response: {},
-				error: false,
-			};
+				return {
+					response: {},
+					error: false,
+				};
+			} finally {
+				yield unlockSurveyTrigger( triggerID );
```

Comment:
This ensures the lock is always released regardless of when the return happens without duplicating throughout.